LACP support

ydirson opened the pull request:
NOTE: [submitted upstream](https://github.com/xenserver/host-installer/pull/37)

This branch deals with the installer side of things, firstboot data needs to be handled in XAPI, for which we depend on https://github.com/xapi-project/xen-api/pull/4333:
- configuration from cmdline (covers fetching answerfile and the next items)
- configuration from answerfile
- configuration UI for netinstall
- configuration UI of mgmt interface, writing data for firstboot

Known limitations we may want to address:
- bad cmdline config could be reported cleanly (we currently get a python backtrace, which gets garbled by the TUI)
- does not deal with DebStyleInterface, ie. does not support Debian-based installer; shouldn't that code be simply be nuked ?
- limitations that could be addressed by moving bond creation/management to a preliminar "advanced networking" step similar to "advanced storage" (adding a new F key for this to initial screen may look bad, though):
  - bond interface is not created immediately, so appears in the interface list as "[no link]". Maybe this can be dealt with, by first creating `ifcfg-*` files without an IP config.
  - cannot change our mind once the eth* to create the bond have been made


This PR explicitly does not deal with:
- creating more than 1 bond
- system upgrade parsing during upgrade (get mgmt-iface config in `ExistingInstallation.readSettings()`), to which `network_db` seems to be unable to report about bonding (but the impact seems to be limited to default values).  The system's bonds are already preserved on upgrade, though.

What the screens currently look like:

![lacp-0-ifaces](https://user-images.githubusercontent.com/1190950/212967721-65b1966d-fe38-48b5-aac8-7efdaf79e032.png)
![lacp-1-create](https://user-images.githubusercontent.com/1190950/212967753-83d9da62-f9f6-4041-bb46-94854207aab0.png)
![lacp-2-confirm](https://user-images.githubusercontent.com/1190950/212967775-2986f3d7-d284-4dc7-b8c0-7b105d6bcd8f.png)
![lacp-3-ifaces](https://user-images.githubusercontent.com/1190950/214070032-87340d94-b620-4ac1-a47d-cc36ab4b9c53.png)

stormi commented:
Thanks for this contribution. Can you give me examples of a name that matches correctly and the one you saw did not match and caused the bug?

stormi commented:
Ping?

nagaozen commented:
Sorry, I was doing this in the office, but had to go forward with another solution. I can't rollback to check the details. The installer is just broken the way it's now. I followed the code to find the bug and prevent future users to end up like me. It's a minor change and the bug is probably introduced by the line 409 for the reasons explained on lines 407 and 408. It's a greedy approach, so I just took a more conservative path, instead of admitting that "mdadm may append an _ followed by a number", I'm taking into account that it might be the case, and update the name if its the case, but keeping the original name otherwise. Here's the error:

![WP_20190611_16_24_02_Pro](https://user-images.githubusercontent.com/22817/59888685-48aea900-939f-11e9-9286-69f9fee7f5a2.jpg)

stormi commented:
I trust your change, but I would like to be able to test the regexp myself at least once on the string that made the installer fail, so if you can get provide it that would greatly help.

stormi commented:
There's probably a way to extract that string from the host even if it's running another solution, since it's hardware-related.

nagaozen commented:
You may want to try _exempli gratia_  `--name=company_namespace` or `--name=alpha_n0`

stormi commented:
I need to understand the context more to be able to try to contribute your patch upstream. 
* In what situations may device names have more than numbers after `_`?
* Are those custom-named devices? What kind of devices? What's the use case?

With your patch, you avoid the error, but you don't adapt the regexp to match those devices (which should probably be preferred). Why this choice?

I know this may look like a lot of discussion for a small patch, but I really need to understand the context and the outcome clearly before merging it, especially since I did not write the code in the first place.

benjamreis committed:
tui.network: make lentry() and iface_details() more accessible

We'll call them from upcoming lacp_bond_ui().

benjamreis committed:
writeRHStyleInterface: split writeIface()

This will allow for more concise logic.

ydirson committed:
doc: make it clear where we expect URLs

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
doc: fix description of --network_device default

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
init/configureNetworking: improve readability

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
doc: make --network_device no less visible than its deprecated counterpart

Don't repeat all parameters on the deprecated version, as we're going to
add more options.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
ifup: remove extraneous assert

This check was already not very useful, as if an eth device is not in
getNetifList(), the ifup command will just fail properly.

But in the case of a bond interface, the bond device will not even exist
before running ifup.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
getNetifList: remove eth hardcoding in interface sorting

This is a prerequisite to allowing more than just eth*

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
netutil.scanConfiguration: rework for extensibility

Less complexity for the nominal case, and makes it easier to add different
non-nominal cases.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
netutil.scanConfiguration: allow reuse of commandline config

Bonding configuration created from commandline cannot be properly
passed to the UI, which needs it both for netinstall and
host-management, so we have to re-discover the existing configuration
to present it to the user.  This includes identifying the configured
bonding interface, and filtering out their member interfaces from
available choices.

This is not very satisfying, and we may want to improve the overall
installer architecture to make this better in the future, but it would be
way out of scope here.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

benjamreis committed:
configureNetworking: write bond configuration to firstboot.d

ydirson committed:
tui: let user configure a bonding interface interactively

Works at selectNetif level, and as such applies both to installer and
host-management networt configs.

benjamreis commented on netutil.py:
```diff
@@ -20,10 +20,20 @@ def __init__(self, nic_dict):
+        nic.bond_mode = bond_mode
+        nic.bond_members = bond_members
+        return nic
+
```
You don't use `cls` why not make a `@staticmethod` then?

ydirson replied:
right, but this is going away anyawy

benjamreis commented:
https://github.com/xapi-project/xen-api/pull/4333

ydirson committed:
answerfile: improve handling of admin-interface selection (xenserver#33)

ydirson committed:
init: add support for LACP bonding in --network_device (xcp-ng/xcp#350)

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

stormi commented on netutil.py:
```diff
@@ -45,16 +45,46 @@ def scanConfiguration():
+            lines = fd.readlines()
+            assert len(lines) == 1
+            if lines[0].strip() != "802.3ad 4":
+                logger.log("scanConfiguration: wrong bonding mode %r" % lines[0])
```
Would it be useful to specify in the log that the expected bonding mode is LACP?

stormi commented on tui/network.py:
```diff
@@ -159,6 +159,44 @@ def iface_details(context, conf):
     tui.screen.popHelpLine()
     return True
 
+def lacp_bond_ui(conf):
```
Next time you test it, could you take screenshots of the UI?

ydirson replied:
ok, when it's ready enough ;)

ydirson committed:
tui: don't hide NIC selection when there is only one

Making things visible is good (esp. here it allows to check interface
details), especially once we add the ability to create a bond
interface.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
NetInterface: remove dead code

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>
(cherry picked from commit 50ecc56bc819f05cb4bc26d6df630de30d01f8de)

ydirson committed:
net*: cleanups around vlan-interface

Using 1 name for 2 concepts is a bad idea, especially in a single function.

Also nukes one line of dead code.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
writeIface: use "with" statement

Always safer, and will facilitate further splitting

ydirson committed:
writeIface: split writing IP configuration for reuse

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
answerfile: LACP bonding support

ydirson committed:
answerfile: remove useless statement

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson commented:
Working on answerfile shows exacerbated need to pass a NIC object to NetInterface ctor (instead of "just hwaddr" initially, getting to which we now add bonding parameters).  The only call sites for which conversion is still not obvious to me are the ones in `product.py`, dealing with the upgrade process.

ydirson committed:
NetInterface: pass a NIC not just a hwaddr to ctor

For bonding support we'll add more fields to NIC, that NetInterface needs
to know about.

ydirson committed:
WIP NIC,NetInterface: add support for bonding interfaces

FIXME:
- maybe support DebStyleInterface ?  how would we test ?
